### PR TITLE
Using constructor property promotion for better types

### PR DESCRIPTION
**Pull request by mamazu**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| License | MIT

#### What's in this PR?

Using constructor property promotion for exceptions and fixing the type issues.

#### Why?

* Less Code duplication
    * Types only need to be defined once
    * Less code for assigning properties

**Isn't this a BC break?**

No, because the properties are not accessible outside or are private and their only write access is already type protected.

#### Example Usage

Before:
```php
/** @var string */
private $test;

/** @param string $test */
public function __construct($test) {
    $this->test = $test;
}
```

After:
```php
/** @param string $test */
public function __construct(private $test) {
}
```

### CODE REVIEW

File: src/Sulu/Component/Rest/Exception/ConstraintViolationException.php [hunk 1]
```diff
@@ -18,36 +18,27 @@ class ConstraintViolationException extends RestException
...
-    {
-        $this->field = $field;
-        $this->entity = $entity;
-        $this->type = $type;
-
+    public function __construct(
+        protected $field,
+        protected $entity,
+        protected $type,
+        protected $code
```
Comment (alexander-schranz) [inline src/Sulu/Component/Rest/Exception/ConstraintViolationException.php]:
```suggestion
       $code
```

Code should just given to the parent

**Review by alexander-schranz — CHANGES_REQUESTED** [review]

File: src/Sulu/Component/Rest/Exception/ConstraintViolationException.php [hunk 2]
```diff
@@ -18,36 +18,27 @@ class ConstraintViolationException extends RestException
...
+    ) {
         parent::__construct(
-            \sprintf('%s constraint for field "%s" of entity "%s" violated', \ucfirst($type), $field, $entity),
-            $code
+            \sprintf('%s constraint for field "%s" of entity "%s" violated', \ucfirst($this->type), $this->field, $this->entity),
+            $this->code
```
Comment (alexander-schranz) [inline src/Sulu/Component/Rest/Exception/ConstraintViolationException.php]:
```suggestion
            $code
```

**Review by alexander-schranz — APPROVED** [review]

File: src/Sulu/Bundle/MarkupBundle/Tag/TagNotFoundException.php [hunk 3]
```diff
@@ -16,33 +16,19 @@
...
-
-    /**
-     * @var string
-     */
-    private $type;
-
     /**
      * @param string $namespace
      * @param string $tagName
      * @param int $type
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MarkupBundle/Tag/TagNotFoundException.php]: @mamazu not sure but I think this should be `string` as it comes from this interface: https://github.com/sulu/sulu/blob/08a40b7de9c1b3f9a6c42edc928ee8e4ff9ea97c/src/Sulu/Bundle/MarkupBundle/Tag/TagRegistryInterface.php#L23

Comment (alexander-schranz) [inline src/Sulu/Bundle/MarkupBundle/Tag/TagNotFoundException.php]:
```suggestion
     * @param string $type
```

File: src/Sulu/Bundle/MarkupBundle/Tag/TagNotFoundException.php [hunk 4]
```diff
@@ -58,7 +44,7 @@ public function getTagName()
     /**
      * Returns type of content.
      *
-     * @return string
+     * @return int
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MarkupBundle/Tag/TagNotFoundException.php]:
```suggestion
     * @return string
```

**Comment by mamazu** [comment]
Your changes all make sense. I'd say go ahead and merge it if there isn't anything else wrong.

**Comment by mamazu** [comment]
@alexander-schranz Can you have a look at this again? I have updated the branch and now it only fails in the same places the 2.5 Branch also fails.

File: src/Sulu/Component/Content/Exception/UnexpectedPropertyType.php [hunk 5]
```diff
@@ -16,35 +16,19 @@
...
+        parent::__construct(
+            \sprintf('Property "%s" is unexpected in content type "%s"', $property->getName(), \get_class($contentType))
+        );
     }
 
-    /**
-     * @return PropertyInterface
-     */
-    public function getProperty()
+    public function getProperty(): PropertyInterface
```
Comment (alexander-schranz) [inline src/Sulu/Component/Content/Exception/UnexpectedPropertyType.php]: For BC need to stay as PHPDoc as this method and this class is not final

File: src/Sulu/Component/Content/Exception/UnexpectedPropertyType.php [hunk 6]
```diff
@@ -16,35 +16,19 @@
...
     {
         return $this->property;
     }
 
-    /**
-     * @return ContentTypeInterface
-     */
-    public function getContentType()
+    public function getContentType(): ContentTypeInterface
```
Comment (alexander-schranz) [inline src/Sulu/Component/Content/Exception/UnexpectedPropertyType.php]: For BC need to stay as PHPDoc as this method and this class is not final

**Review by alexander-schranz — CHANGES_REQUESTED** [review]

**Comment by mamazu** [comment]
Yes you are correct. I have removed them.

**Review by alexander-schranz — APPROVED** [review]

**Comment by alexander-schranz** [comment]
@mamazu thank you!

**Pull request merged**